### PR TITLE
Add API generator for getting custom parsing results.

### PR DESCRIPTION
**Pull request by ivanisidrowu**
## Modified

- Remove the limitation of the reified type of the API.
- Add API generator for getting custom parsing results.

## Examples

The examples below show how to annotate classes and what's the generated code looks like.

### Annotated Class

```kotlin
@RssTag(name = "channel")
data class RssData(
    val title: String?,
    val textInput: MyTextInput?,
    @RssTag(name = "item")
    val list: List<RssItem>,
    val cloud: Cloud?,
    @RssTag(name = "category")
    val categories: List<Category>,
): Serializable

@RssTag(name = "textInput")
data class MyTextInput(
    val title: String?,
    val name: String?
): Serializable

@RssTag(name = "item")
data class RssItem(
    val title: String?,
    val author: String?,
): Serializable

@RssTag(name = "cloud")
data class Cloud(
    @RssAttribute
    val domain: String?,
    @RssAttribute(name = "port")
    val portNum: Int?,
    @RssAttribute
    val path: String?
): Serializable

@RssTag(name = "category")
data class Category(
    @RssAttribute
    val domain: String?
)
```

### Generated Code

```kotlin
object RssDataReader {

  inline fun read(url: String, config: KtRssReaderConfig.() -> Unit = {}): RssData =
      Reader.read<RssData>(url = url, customParser = { xml -> RssDataParser.parse(xml) }, config)

  inline fun flowRead(url: String, crossinline config: KtRssReaderConfig.() -> Unit = {}):
      Flow<RssData> = Reader.flowRead<RssData>(url = url, customParser = { xml ->
      RssDataParser.parse(xml) }, config)

  suspend inline fun coRead(url: String, crossinline config: KtRssReaderConfig.() -> Unit = {}):
      RssData = Reader.coRead<RssData>(url = url, customParser = { xml -> RssDataParser.parse(xml)
      }, config)
      
}
```

## TODO
- Add tests for custom parsers.
- Add examples of custom parsers.
- Modify README.

### CODE REVIEW

File: ktRssReader/src/main/java/tw/ktrssreader/utils/TopLevelFunctions.kt [hunk 1]
```diff
@@ -62,18 +62,19 @@ fun <T> ByteArray.convertToObject(): T {
...
     ifGoogle: () -> Any,
-    ifAutoMix: () -> Any
-): R {
+    ifAutoMix: () -> Any,
+): R? {
     val clazz = T::class.java
     // Do NOT change this order
     return when {
         AutoMixChannel::class.java.isAssignableFrom(clazz) -> ifAutoMix()
+        RssStandardChannel::class.java.isAssignableFrom(clazz) -> ifRssStandard()
```
Comment (dev-weiqi) [inline ktRssReader/src/main/java/tw/ktrssreader/utils/TopLevelFunctions.kt:75]: 這個順序應該要在最後面
因為判斷的時候進來如果是 google channel 他的 Interface 屬於 RssStandardChannel

File: processor/src/main/java/tw/ktrssreader/processor/RssAnnotationProcessor.kt [hunk 2]
```diff
@@ -58,11 +62,19 @@ class RssAnnotationProcessor : AbstractProcessor() {
         p1?.getElementsAnnotatedWith(RssTag::class.java)?.forEach {
             if (it?.kind == ElementKind.CLASS) {
                 val rssTag = it.getAnnotation(RssTag::class.java)
+                val isRoot = rssTag?.name == CHANNEL
                 ParserGenerator(
                     element = it,
                     isRoot = rssTag?.name == CHANNEL,
```
Comment (dev-weiqi) [inline processor/src/main/java/tw/ktrssreader/processor/RssAnnotationProcessor.kt:68]: 這邊可以用上面的 isRoot

**Review by dev-weiqi — COMMENTED** [review]
Done

File: ktRssReader/src/main/java/tw/ktrssreader/utils/TopLevelFunctions.kt (hunk 1 above)
Comment (ivanisidrowu) [reply ktRssReader/src/main/java/tw/ktrssreader/utils/TopLevelFunctions.kt]: done

File: processor/src/main/java/tw/ktrssreader/processor/RssAnnotationProcessor.kt (hunk 2 above)
Comment (ivanisidrowu) [reply processor/src/main/java/tw/ktrssreader/processor/RssAnnotationProcessor.kt]: done

**Review by dev-weiqi — APPROVED** [review]
👍🏻

**Review by HsienDaDa — APPROVED** [review]

**Pull request merged**